hpx/lcos/dist_object.hpp

```diff
@@ -0,0 +1,566 @@
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+/// \file lcos/dist_object.hpp
+
+#ifndef HPX_LCOS_DIST_OBJECT_HPP
+#define HPX_LCOS_DIST_OBJECT_HPP
+
+#include <hpx/include/components.hpp>
+#include <hpx/include/lcos.hpp>
+#include <hpx/lcos/barrier.hpp>
+#include <hpx/lcos/local/spinlock.hpp>
+#include <hpx/pp/cat.hpp>
+#include <hpx/runtime/actions/component_action.hpp>
+#include <hpx/util/assert.hpp>
+#include <hpx/util/bind.hpp>
+
+#include <chrono>
+#include <cstddef>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+/// \cond NOINTERNAL
+namespace hpx { namespace lcos { namespace server {
+    template <typename T>
+    class dist_object_part
+      : public hpx::components::locking_hook<
+            hpx::components::component_base<dist_object_part<T>>>
+    {
+    public:
+        typedef T data_type;
+        dist_object_part() {}
+
+        dist_object_part(data_type const& data)
+          : data_(data)
+        {
+        }
+
+        dist_object_part(data_type&& data)
+          : data_(std::move(data))
+        {
+        }
+
+        data_type& operator*()
+        {
+            return data_;
+        }
+
+        data_type const& operator*() const
+        {
+            return data_;
+        }
+
+        data_type const* operator->() const
+        {
+            return &data_;
+        }
+
+        data_type* operator->()
+        {
+            return &data_;
+        }
+
+        data_type fetch() const
+        {
+            return data_;
+        }
+
+        HPX_DEFINE_COMPONENT_ACTION(dist_object_part, fetch);
+
+    private:
+        data_type data_;
+    };
+
+    template <typename T>
+    class dist_object_part<T&>
+      : public hpx::components::locking_hook<
+            hpx::components::component_base<dist_object_part<T&>>>
+    {
+    public:
+        typedef T& data_type;
+        dist_object_part() {}
+
+        dist_object_part(data_type data)
+          : data_(data)
+        {
+        }
+
+        data_type operator*()
+        {
+            return data_;
+        }
+
+        data_type operator*() const
+        {
+            return data_;
+        }
+
+        T const* operator->() const
+        {
+            return data_;
+        }
+
+        T* operator->()
+        {
+            return data_;
+        }
+
+        T fetch() const
+        {
+            return data_;
+        }
+
+        HPX_DEFINE_COMPONENT_ACTION(dist_object_part, fetch);
+
+    private:
+        data_type data_;
+    };
+}}}
+
+#define REGISTER_DIST_OBJECT_PART_DECLARATION(type)                            \
+    HPX_REGISTER_ACTION_DECLARATION(                                           \
+        hpx::lcos::server::dist_object_part<type>::fetch_action,               \
+        HPX_PP_CAT(__dist_object_part_fetch_action_, type));
+
+/**/
+
+#define REGISTER_DIST_OBJECT_PART(type)                                        \
```

Comment:
Also naming here, I'd prefer (even though it's a bit long winded)
`REGISTER_DISTRIBUTED_OBJECT_PART` ?
and so on for other macros

Reply:
Updated as well. Everything is now named `distributed_object`, including file name.